test/blackbox/common/BlackboxTestsDeadlineQos.cpp

```diff
@@ -309,6 +334,169 @@ TEST_P(DeadlineQos, KeyedTopicBestEffortReaderVolatileWriterSetDeadline)
+
+    eprosima::fastdds::dds::DataWriter& native = writer.get_native_writer();
+
+    DeadlineQos::set_deadline_seconds(native, 0.15); // Update 0 -> finite: counters must reset, no new warning
+
+    eprosima::fastdds::dds::OfferedDeadlineMissedStatus st{};
+    std::this_thread::sleep_for(std::chrono::milliseconds(150));
+    ASSERT_EQ(native.get_offered_deadline_missed_status(st), eprosima::fastdds::dds::RETCODE_OK);
+    EXPECT_EQ(st.total_count, 0u);
+    EXPECT_EQ(st.total_count_change, 0u);
+    EXPECT_EQ(consumer_ptr->count(), prev) << "No new warning when moving from 0 -> finite";
+
+    DeadlineQos::set_deadline_seconds(native, 0.0); // Update finite -> 0: exactly one new warning, counters jump to max
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    ASSERT_EQ(native.get_offered_deadline_missed_status(st), eprosima::fastdds::dds::RETCODE_OK);
+    EXPECT_EQ(consumer_ptr->count(), prev + 1) << "Exactly one extra warning on finite -> 0";
+    EXPECT_EQ(st.total_count, std::numeric_limits<uint32_t>::max());
+    EXPECT_EQ(st.total_count_change, std::numeric_limits<uint32_t>::max());
+
+    fastlog::Log::ClearConsumers();
+}
+
+/**
+ * Testing Redmine issue #23289.
+ * Regression test for the deadline period change at runtime.
+ * Creating a DataWriter with a small deadline so missed deadlines are registered.
+ * Changing deadline period to a large value to infer that there are no missed deadlines anymore.
+ * Checking total count and count change when changing the deadline
+ */
+TEST_P(DeadlineQos, ChangeDeadlineAtRuntime)
+{
+    // Writer/Reader setup (same pattern as your other tests)
+    PubSubWriter<KeyedHelloWorldPubSubType> writer(TEST_TOPIC_NAME);
+    PubSubReader<KeyedHelloWorldPubSubType> reader(TEST_TOPIC_NAME);
+
+    writer.durability_kind(eprosima::fastdds::dds::VOLATILE_DURABILITY_QOS);
+    writer.reliability(eprosima::fastdds::dds::RELIABLE_RELIABILITY_QOS);
+    reader.reliability(eprosima::fastdds::dds::RELIABLE_RELIABILITY_QOS);
+
+    writer.deadline_period(0.01).init();
+    reader.init();
+    ASSERT_TRUE(writer.isInitialized());
+    ASSERT_TRUE(reader.isInitialized());
+    writer.wait_discovery();
+    reader.wait_discovery();
+
+    eprosima::fastdds::dds::DataWriter& native = writer.get_native_writer();
+
+    // Arm the per-instance timer
+    KeyedHelloWorld sample;
+    sample.key(1);
+    ASSERT_EQ(native.write(&sample), eprosima::fastdds::dds::RETCODE_OK);
+
+    // Let several deadlines pass
+    eprosima::fastdds::dds::OfferedDeadlineMissedStatus st{};
+    std::this_thread::sleep_for(std::chrono::milliseconds(150));
+    ASSERT_EQ(native.get_offered_deadline_missed_status(st), eprosima::fastdds::dds::RETCODE_OK);
+    EXPECT_GT(st.total_count, 0u);
+    EXPECT_GT(st.total_count_change, 0u);
+
+    // Switch to a large deadline
+    DeadlineQos::set_deadline_seconds(native, 5000.0);
+
+    // Immediately after the change, there can be one last miss in flight
```

Comment:
Maybe it would be better to make the deadline period greater and be stricter (not allow for any single extra deadline misses).